source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -156,14 +174,43 @@ Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::ResponseHeaderMap&
   } else {
     headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.GrpcWebProto);
   }
-  return Http::FilterHeadersStatus::Continue;
+
+  if (end_stream || isValidResponseHeaders(headers, end_stream)) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  ENVOY_LOG(debug, "received invalid response headers since the upstream response or local reply "
```

Comment:
If we think it's invalid for the local reply to not send gRPC web, I'd suggest that sendLocalReply track both is_grpc and is_grpc_web.  Then this filter need not transform. WDYT?

Reply:
Yes, I think that will be one of the solution. I discussed this offline with @snowp, since we need to transform bad upstream response as well, probably this is enough to handle it here. Either way I'm good.

Reply:
Accommodated.